File: Pinball App 2/Pinball App 2/practice/PracticeDisplayTitles.swift
import Foundation

func practiceDisplayTitleForGames(_ games: [PinballGame]) -> String {
    let baseCandidates = games
        .map(\.name)
        .map(practiceNormalizedDisplayBaseTitle)
        .filter { !$0.isEmpty }
    if let best = practicePreferredDisplayTitleCandidate(from: baseCandidates) {
        return best
    }

    let opdbCandidates = games
        .compactMap(\.opdbName)
        .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
        .filter { !$0.isEmpty }
        .map(practiceNormalizedDisplayBaseTitle)
        .filter { !$0.isEmpty }
    if let best = practicePreferredDisplayTitleCandidate(from: opdbCandidates) {
        return best
    }

    return games.map(\.name).min(by: practiceDisplayTitleOrder) ?? games.first?.name ?? "Unknown Game"
}

func practiceDisplayTitle(for canonicalGameID: String, in games: [PinballGame]) -> String? {
    let trimmed = canonicalGameID.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmed.isEmpty else { return nil }
    let grouped = games.filter { $0.canonicalPracticeKey == trimmed }
    guard !grouped.isEmpty else { return nil }
    return practiceDisplayTitleForGames(grouped)
}

private func practiceNormalizedDisplayBaseTitle(_ raw: String) -> String {
    var current = raw.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !current.isEmpty else { return "" }

    let pattern = #"\s*\([^()]*\)\s*$"#
    while let range = current.range(of: pattern, options: .regularExpression) {
        let candidate = current.replacingCharacters(in: range, with: "")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        guard !candidate.isEmpty else { break }
        current = candidate
    }

    return current
}

private func practicePreferredDisplayTitleCandidate(from candidates: [String]) -> String? {
    let normalized = candidates
        .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
        .filter { !$0.isEmpty }
    guard !normalized.isEmpty else { return nil }

    let counts = Dictionary(grouping: normalized, by: {
        $0.folding(options: [.diacriticInsensitive], locale: .current).lowercased()
    })
    let ranked = counts.values.compactMap { group -> String? in
        group.min(by: practiceDisplayTitleOrder)
    }

    return ranked.max { lhs, rhs in
        let lhsCount = counts[lhs.folding(options: [.diacriticInsensitive], locale: .current).lowercased()]?.count ?? 0
        let rhsCount = counts[rhs.folding(options: [.diacriticInsensitive], locale: .current).lowercased()]?.count ?? 0
        if lhsCount != rhsCount { return lhsCount < rhsCount }
        return !practiceDisplayTitleOrder(lhs, rhs)
    }
}

private func practiceDisplayTitleOrder(_ lhs: String, _ rhs: String) -> Bool {
    let lhsTrimmed = lhs.trimmingCharacters(in: .whitespacesAndNewlines)
    let rhsTrimmed = rhs.trimmingCharacters(in: .whitespacesAndNewlines)
    if lhsTrimmed.count != rhsTrimmed.count {
        return lhsTrimmed.count < rhsTrimmed.count
    }
    return lhsTrimmed.localizedCaseInsensitiveCompare(rhsTrimmed) == .orderedAscending
}

File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/practice/PracticeDisplayTitles.kt
package com.pillyliu.pinprofandroid.practice

import com.pillyliu.pinprofandroid.library.PinballGame
import java.util.Locale

internal fun practiceDisplayTitleForGames(games: List<PinballGame>): String {
    val baseCandidates = games
        .map { practiceNormalizedDisplayBaseTitle(it.name) }
        .filter { it.isNotBlank() }
    practicePreferredDisplayTitleCandidate(baseCandidates)?.let { return it }

    val opdbCandidates = games
        .mapNotNull { it.opdbName?.trim()?.takeIf(String::isNotBlank) }
        .map(::practiceNormalizedDisplayBaseTitle)
        .filter { it.isNotBlank() }
    practicePreferredDisplayTitleCandidate(opdbCandidates)?.let { return it }

    return games.map { it.name }.minWithOrNull(practiceDisplayTitleComparator()) ?: "Unknown Game"
}

internal fun practiceDisplayTitleForKey(
    canonicalGameId: String,
    games: List<PinballGame>,
): String? {
    val trimmed = canonicalGameId.trim()
    if (trimmed.isBlank()) return null
    val grouped = games.filter { it.practiceKey == trimmed }
    if (grouped.isEmpty()) return null
    return practiceDisplayTitleForGames(grouped)
}

private fun practiceNormalizedDisplayBaseTitle(raw: String): String {
    var current = raw.trim()
    if (current.isBlank()) return ""

    val pattern = Regex("\\s*\\([^()]*\\)\\s*$")
    while (true) {
        val replaced = pattern.replace(current, "").trim()
        if (replaced.isBlank() || replaced == current) break
        current = replaced
    }

    return current
}

private fun practicePreferredDisplayTitleCandidate(candidates: List<String>): String? {
    val normalized = candidates.map { it.trim() }.filter { it.isNotBlank() }
    if (normalized.isEmpty()) return null

    val grouped = normalized.groupBy { it.lowercase(Locale.US) }
    return grouped.values
        .mapNotNull { group -> group.minWithOrNull(practiceDisplayTitleComparator()) }
        .maxWithOrNull(
            compareBy<String> { grouped[it.lowercase(Locale.US)]?.size ?: 0 }
                .thenByDescending { -it.length }
                .thenByDescending { it.lowercase(Locale.US) }
        )
}

private fun practiceDisplayTitleComparator(): Comparator<String> {
    return compareBy<String> { it.trim().length }.thenBy { it.lowercase(Locale.US) }
}
